Kernel/ACPI/ACPIStaticParser.cpp

```diff
@@ -51,45 +52,27 @@ ACPI_RAW::SDTHeader* ACPIStaticParser::find_table(const char* sig)
     return nullptr;
 }
 
-ACPI_RAW::SDTHeader& ACPIStaticParser::find_fadt()
-{
-    kprintf("ACPI: Searching for the Fixed ACPI Data Table\n");
-    for (auto* physical_sdt_ptr : m_main_sdt->get_sdt_pointers()) {
-        auto region = MM.allocate_kernel_region((PAGE_SIZE * 2), "ACPI Static Parser AML Tables Finding", Region::Access::Read);
-        mmap_region(*region, PhysicalAddress((u32)physical_sdt_ptr & PAGE_MASK));
-        ACPI_RAW::SDTHeader* sdt = (ACPI_RAW::SDTHeader*)(region->vaddr().get() + ((u32)physical_sdt_ptr & (~PAGE_MASK)));
-#ifdef ACPI_DEBUG
-        dbgprintf("ACPI: Examining Table @ P 0x%x\n", physical_sdt_ptr);
-#endif
-        if (!strncmp(sdt->sig, "FACP", 4)) {
-            kprintf("ACPI: Found FADT Table @ P 0x%x, registering\n", physical_sdt_ptr);
-            return *physical_sdt_ptr;
-        }
-    }
-    ASSERT_NOT_REACHED(); // If we reached this point, there's something very incorrect with the system!
-}
-
 void ACPIStaticParser::init_fadt()
 {
     kprintf("ACPI: Initializing Fixed ACPI data\n");
-    ACPI_RAW::SDTHeader& physical_fadt = find_fadt();
+    kprintf("ACPI: Searching for the Fixed ACPI Data Table\n");
+    ASSERT(find_table("FACP") != nullptr);
+    auto* fadt_ptr = find_table("FACP");
 
-    auto checkup_region = MM.allocate_kernel_region((PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
+    auto checkup_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)(fadt_ptr))), (PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
 #ifdef ACPI_DEBUG
     dbgprintf("ACPI: Checking FADT Length to choose the correct mapping size\n");
 #endif
-    mmap_region(*checkup_region, PhysicalAddress((u32)(&physical_fadt) & PAGE_MASK));
 
-    ACPI_RAW::SDTHeader* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().get() + ((u32)(&physical_fadt) & (~PAGE_MASK)));
+    ACPI_RAW::SDTHeader* sdt = (ACPI_RAW::SDTHeader*)(create_vm_address(checkup_region->vaddr(), PhysicalAddress((u32)(fadt_ptr)))).as_ptr();
 #ifdef ACPI_DEBUG
-    dbgprintf("ACPI: FADT @ V 0x%x, P 0x%x\n", sdt, &physical_fadt);
+    dbgprintf("ACPI: FADT @ V 0x%x, P 0x%x\n", sdt, fadt_ptr);
 #endif
     u32 length = sdt->length;
     kprintf("ACPI: Fixed ACPI data, Revision %u\n", sdt->revision);
 
-    auto fadt_region = MM.allocate_kernel_region(PAGE_ROUND_UP(length) + PAGE_SIZE, "ACPI Static Parser", Region::Access::Read);
-    mmap_region(*fadt_region, PhysicalAddress((u32)(&physical_fadt) & PAGE_MASK));
-    m_fadt = OwnPtr<ACPI::FixedACPIData>(new ACPI::FixedACPIData(*((ACPI_RAW::FADT*)(fadt_region->vaddr().get() + ((u32)(&physical_fadt) & (~PAGE_MASK))))));
+    auto fadt_region = MM.allocate_kernel_region(PhysicalAddress(page_base_of((u32)(fadt_ptr))), PAGE_ROUND_UP(length) + PAGE_SIZE, "ACPI Static Parser", Region::Access::Read);
+    m_fadt = OwnPtr<ACPI::FixedACPIData>(new ACPI::FixedACPIData(*((ACPI_RAW::FADT*)create_vm_address(fadt_region->vaddr(), PhysicalAddress((u32)(fadt_ptr))).as_ptr())));
```

Comment:
```suggestion
    m_fadt = make<ACPI::FixedACPIData>(*(ACPI_RAW::FADT*)fadt_region->vaddr().page_base().offset((u32)fadt_ptr).as_ptr());
```

Reply:
```fadt_region->vaddr().page_base().offset((u32)fadt_ptr).as_ptr()```
is wrong, since it won't trim the upper bits as I mentioned above.
But, using make<> is a nice suggestion :)